### PR TITLE
Add payment refund handler

### PR DESCRIPTION
**Pull request by felixbrucker**
This PR adds a payment refund handler to the paypal payment method. This allows payments created with order transaction captures to be refunded through Shopware.

### CODE REVIEW

File: src/Checkout/PayPalOrderTransactionCaptureService.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        $data = [
+            'id' => $orderTransactionCaptureId,
+            'externalReference' => $paypalResourceId,
+            'customFields' => [
+                SwagPayPal::ORDER_TRANSACTION_CAPTURE_CUSTOM_FIELDS_PAYPAL_RESOURCE_TYPE => $paypalResourceType,
+            ],
+        ];
+        $this->orderTransactionCaptureRepository->update([$data], $context);
+    }
+}
```
Comment (fixpunkt) [inline src/Checkout/PayPalOrderTransactionCaptureService.php]: Missing LF

File: src/PaymentsApi/Administration/PayPalPaymentController.php [hunk 2]
```diff
@@ -263,22 +283,49 @@ public function capturePayment(
...
+        try {
+            switch ($resourceType) {
+                case RelatedResource::AUTHORIZE:
+                    $captureResponse = $this->authorizationResource->capture(
+                        $resourceId,
+                        $capture,
+                        $this->getSalesChannelIdByOrderId($orderId, $context)
+                    );
+
+                    break;
```
Comment (fixpunkt) [inline src/PaymentsApi/Administration/PayPalPaymentController.php]:
```suggestion
                    break;

```

File: src/PaymentsApi/Administration/PayPalPaymentController.php [hunk 3]
```diff
@@ -263,22 +283,49 @@ public function capturePayment(
...
+                case RelatedResource::ORDER:
+                    $salesChannelId = $this->getSalesChannelIdByOrderId($orderId, $context);
+                    $captureResponse = $this->ordersResource->capture($resourceId, $capture, $salesChannelId);
+
+                    break;
```
Comment (fixpunkt) [inline src/PaymentsApi/Administration/PayPalPaymentController.php]:
```suggestion
                    break;

```

File: src/Refund/PayPalRefundHandler.php [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+            $salesChannelId = $orderRefund->getOrder()->getSalesChannelId();
+            switch ($resourceType) {
+                case RelatedResource::SALE:
+                    $refundResponse = $this->saleResource->refund(
+                        $resourceIdToRefund,
+                        $refund,
+                        $salesChannelId
+                    );
+                    break;
```
Comment (fixpunkt) [inline src/Refund/PayPalRefundHandler.php:96]:
```suggestion
                    break;

```

File: src/Refund/PayPalRefundHandler.php [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                case RelatedResource::CAPTURE:
+                    $refundResponse = $this->captureResource->refund(
+                        $resourceIdToRefund,
+                        $refund,
+                        $salesChannelId
+                    );
+                    break;
```
Comment (fixpunkt) [inline src/Refund/PayPalRefundHandler.php:103]:
```suggestion
                    break;

```

File: src/Refund/PayPalRefundHandler.php [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        } catch (\Throwable $e) {
+            throw new PaymentRefundProcessException($orderRefundTransactionId, $e->getMessage());
+        }
+    }
+
+    private function createRefund(OrderRefundEntity $orderRefund): Refund
+    {
+        $refundAmount = $this->priceFormatter->formatPrice($orderRefund->getAmount());
+        $currency = $orderRefund->getOrder()->getCurrency()->getShortName();
+        $invoiceNumber = ''; // TODO: store in order refund transaction?
```
Comment (fixpunkt) [inline src/Refund/PayPalRefundHandler.php]: TODO

**Review by fixpunkt — CHANGES_REQUESTED** [review]

**Review by fixpunkt — CHANGES_REQUESTED** [review]
Re https://github.com/pickware/SwagPayPal/pull/1#discussion_r545254306 and other comments:

If formatting of `case...break` is inconsistent in the PayPal codebase anyway, I'd strongly prefer using 

```php
case 0:
    // do stuff
    break;
case 1:
```

over

```php
case 0:
    // do stuff

    break;
case 1:
```

**Review by fixpunkt — APPROVED** [review]
Very cool!

**Comment by felixbrucker** [comment]
Closing here in favor of https://github.com/shopwareLabs/SwagPayPal/pull/14

**Pull request closed without merging**